src/shadowbox/integration_test/test.sh

```diff
@@ -94,24 +112,42 @@ function client_curl() {
   # Create new shadowbox user.
   # TODO(bemasc): Verify that the server is using the right certificate
   declare -r NEW_USER_JSON=$(client_curl --insecure -X POST https://shadowbox/${SB_API_PREFIX}/access-keys)
-  [[ ${NEW_USER_JSON} == '{"id":'* ]] || exit 1
+  [[ ${NEW_USER_JSON} == '{"id":"0"'* ]] || fail "Fail to create user"
   declare -r SS_USER_ARGUMENTS=$(ss_arguments_for_user $NEW_USER_JSON)
 
+  # Verify that we handle deletions well
+  client_curl --insecure -X POST https://shadowbox/${SB_API_PREFIX}/access-keys > /dev/null
+  client_curl --insecure -X DELETE https://shadowbox/${SB_API_PREFIX}/access-keys/1 > /dev/null
+
   # Start Shadowsocks client and wait for it to be ready
   declare -r LOCAL_SOCKS_PORT=5555
   docker exec -d $CLIENT_CONTAINER \
-    /go/bin/go-shadowsocks2 $SS_USER_ARGUMENTS -socks :$LOCAL_SOCKS_PORT -verbose
+    /go/bin/go-shadowsocks2 $SS_USER_ARGUMENTS -socks :$LOCAL_SOCKS_PORT -verbose \
+    || fail "Could not start shadowsocks client"
   while ! docker exec $CLIENT_CONTAINER nc -z localhost $LOCAL_SOCKS_PORT; do
     sleep 0.1
   done
 
   # Verify we can retrieve the target by IP.
   client_curl -x socks5h://localhost:$LOCAL_SOCKS_PORT $TARGET_IP > $OUTPUT_DIR/actual.html
-  diff $OUTPUT_DIR/actual.html target/index.html
+  diff $OUTPUT_DIR/actual.html target/index.html || fail "Target page by IP does not match"
 
   # Verify we can retrieve the target using the system nameservers.
   client_curl -x socks5h://localhost:$LOCAL_SOCKS_PORT http://target > $OUTPUT_DIR/actual.html
-  diff $OUTPUT_DIR/actual.html target/index.html
+  diff $OUTPUT_DIR/actual.html target/index.html || fail "Target page by hostname does not match"
+
+  # Verify we can't access the page anymore after the key is deleted
+  client_curl --insecure -X DELETE https://shadowbox/${SB_API_PREFIX}/access-keys/0 > /dev/null
+  # Exit code 56 is "Connection reset by peer".
+  client_curl -x socks5h://localhost:$LOCAL_SOCKS_PORT --connect-timeout 1 http://target &> /dev/null \
+    && fail "Deleted access key is still active" || (($? == 56))
+
+  # Verify no errors occurred
+  docker logs $SHADOWBOX_CONTAINER &> $OUTPUT_DIR/logs.txt
```

Comment:
@fortuna I'm not a big fan of this move into whitebox testing. I'm not clear why we would fail despite the behaviour being correct. This is making the tests flaky - just sent https://github.com/Jigsaw-Code/outline-server/pull/299 so that the log will be written to the console.

Reply:
I'm with you: i'd rather we had 100% unit test and integration test coverage. But that's not quite the case. This test captured issues with key deletion, allowing me to more easily fix them. They could have prevented the errors from getting to production if it was in place before.

The presence of errors should be deterministic. I don't know why you think this makes the test flaky. On the other hand, the test does make our releases safer.